// @inrupt/solid-sdk-forms/dist/lib/classes/form-actions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var uuid_1 = require("uuid");
var query_ldflex_1 = require("@solid/query-ldflex");
var data_model_1 = require("@rdfjs/data-model");
var _constants_1 = require("@constants");
var _utils_1 = require("@utils");
/**
 * FormAction class to managament Form Model forms on solid
 */
var FormActions = /** @class */ (function () {
    function FormActions(formModel, formObject) {
        var _this = this;
        this.formModel = formModel;
        this.formObject = formObject;
        /**
         * Retrieve new Form Object
         */
        this.retrieveNewFormObject = function (item, value) {
            var _a;
            _this.formObject = __assign({}, _this.formObject, (_a = {}, _a[item] = value, _a));
            return _this.formObject;
        };
        this.validator = function (field) {
            var _a;
            var updatedField = __assign({}, field, (_a = {}, _a[_constants_1.UI.VALID] = true, _a));
            try {
                var _loop_1 = function (currentValidator) {
                    var _a;
                    if (Object.keys(field).find(function (key) { return key === currentValidator.name; })) {
                        var _b = currentValidator.action(field), valid = _b.valid, errorMessage = _b.errorMessage;
                        if (!valid) {
                            updatedField = __assign({}, updatedField, (_a = {}, _a[_constants_1.UI.VALID] = valid, _a[_constants_1.UI.DEFAULT_ERROR] = errorMessage, _a));
                            return "break";
                        }
                    }
                };
                for (var _i = 0, _b = _utils_1.validator.validators; _i < _b.length; _i++) {
                    var currentValidator = _b[_i];
                    var state_1 = _loop_1(currentValidator);
                    if (state_1 === "break")
                        break;
                }
            }
            catch (error) {
                throw Error(error);
            }
            return updatedField;
        };
        /**
         * Save data intot he pod
         */
        this.saveData = function (customFormObject) { return __awaiter(_this, void 0, void 0, function () {
            var keyFields, keyFields_1, keyFields_1_1, key, currentField, validatedField, isType, predicate, value, podData, updatedValue, parent_1, e_1_1;
            var e_1, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        keyFields = Object.keys(customFormObject || this.formObject);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 18, 19, 24]);
                        keyFields_1 = __asyncValues(keyFields);
                        _b.label = 2;
                    case 2: return [4 /*yield*/, keyFields_1.next()];
                    case 3:
                        if (!(keyFields_1_1 = _b.sent(), !keyFields_1_1.done)) return [3 /*break*/, 17];
                        key = keyFields_1_1.value;
                        currentField = this.formObject[key];
                        validatedField = __assign({}, currentField);
                        if (!currentField) return [3 /*break*/, 16];
                        validatedField = this.validator(currentField);
                        isType = currentField[_constants_1.UI.PROPERTY].includes('type');
                        if (!validatedField[_constants_1.UI.VALID]) return [3 /*break*/, 14];
                        predicate = currentField[_constants_1.UI.PROPERTY];
                        value = currentField.value;
                        podData = void 0;
                        validatedField[_constants_1.UI.VALUE] = isType ? data_model_1.namedNode(currentField.value) : currentField.value;
                        updatedValue = typeof value === 'boolean' ? value.toString() : value;
                        if (!currentField.parent) return [3 /*break*/, 9];
                        podData = isType
                            ? query_ldflex_1.default[currentField.parent[_constants_1.UI.VALUE]].type
                            : query_ldflex_1.default[currentField.parent[_constants_1.UI.VALUE]][predicate];
                        if (!(currentField[_constants_1.UI.OLDVALUE] && currentField[_constants_1.UI.OLDVALUE] !== '')) return [3 /*break*/, 5];
                        return [4 /*yield*/, podData.set(updatedValue)];
                    case 4:
                        _b.sent();
                        return [3 /*break*/, 8];
                    case 5:
                        parent_1 = currentField.parent;
                        return [4 /*yield*/, query_ldflex_1.default[parent_1[_constants_1.UI.BASE]][parent_1[_constants_1.UI.PARENT_PROPERTY]].add(data_model_1.namedNode(parent_1[_constants_1.UI.VALUE]))];
                    case 6:
                        _b.sent();
                        return [4 /*yield*/, podData.add(updatedValue)];
                    case 7:
                        _b.sent();
                        _b.label = 8;
                    case 8: return [3 /*break*/, 13];
                    case 9:
                        podData = isType ? query_ldflex_1.default[currentField[_constants_1.UI.BASE]] : query_ldflex_1.default[currentField[_constants_1.UI.BASE]][predicate];
                        if (!currentField[_constants_1.UI.OLDVALUE]) return [3 /*break*/, 11];
                        return [4 /*yield*/, podData.set(updatedValue)];
                    case 10:
                        _b.sent();
                        return [3 /*break*/, 13];
                    case 11: return [4 /*yield*/, podData.add(updatedValue)];
                    case 12:
                        _b.sent();
                        _b.label = 13;
                    case 13: return [3 /*break*/, 15];
                    case 14: throw new Error('Validation failed');
                    case 15:
                        /**
                         * Update ui:value and  ui:oldValue on formModel and reset formObject
                         */
                        this.updateFieldModel(validatedField[_constants_1.UI.NAME], isType ? validatedField[_constants_1.UI.VALUE] : validatedField.value, validatedField[_constants_1.UI.VALID], validatedField[_constants_1.UI.DEFAULT_ERROR]);
                        _b.label = 16;
                    case 16: return [3 /*break*/, 2];
                    case 17: return [3 /*break*/, 24];
                    case 18:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 24];
                    case 19:
                        _b.trys.push([19, , 22, 23]);
                        if (!(keyFields_1_1 && !keyFields_1_1.done && (_a = keyFields_1.return))) return [3 /*break*/, 21];
                        return [4 /*yield*/, _a.call(keyFields_1)];
                    case 20:
                        _b.sent();
                        _b.label = 21;
                    case 21: return [3 /*break*/, 23];
                    case 22:
                        if (e_1) throw e_1.error;
                        return [7 /*endfinally*/];
                    case 23: return [7 /*endfinally*/];
                    case 24:
                        this.resetFormObject();
                        return [2 /*return*/, this.formModel];
                }
            });
        }); };
        /**
         * Reset FormObject
         */
        this.resetFormObject = function () {
            _this.formObject = {};
        };
        /**
         * Field field object into Form Model
         */
        this.updateFieldModel = function (name, newValue, valid, errorMessage) {
            var _a;
            if (valid === void 0) { valid = true; }
            var partsObject = _this.formModel[_constants_1.UI.PARTS];
            var found = false;
            function findRecursive(name, value, model) {
                var _a, _b, _c, _d;
                for (var fieldKey in model) {
                    if (model[fieldKey][_constants_1.UI.NAME] === name) {
                        model = __assign({}, model, (_a = {}, _a[fieldKey] = __assign({}, model[fieldKey], (_b = {}, _b[_constants_1.UI.VALUE] = newValue, _b[_constants_1.UI.OLDVALUE] = newValue !== '' ? newValue : model[fieldKey][_constants_1.UI.OLDVALUE], _b[_constants_1.UI.VALID] = valid, _b[_constants_1.UI.DEFAULT_ERROR] = errorMessage, _b)), _a));
                        found = true;
                        break;
                    }
                    else if (model[fieldKey][_constants_1.UI.PARTS]) {
                        model = __assign({}, model, (_c = {}, _c[fieldKey] = __assign({}, model[fieldKey], (_d = {}, _d[_constants_1.UI.PARTS] = __assign({}, findRecursive(name, value, model[fieldKey][_constants_1.UI.PARTS])), _d)), _c));
                        if (found) {
                            break;
                        }
                    }
                }
                return model;
            }
            try {
                var updatedModel = findRecursive(name, newValue, partsObject);
                /**
                 * Update private formModel store
                 */
                _this.formModel = __assign({}, _this.formModel, (_a = {}, _a[_constants_1.UI.PARTS] = __assign({}, updatedModel), _a));
                return _this.formModel;
            }
            catch (error) {
                return Error(error);
            }
        };
        /**
         * Delete field into Form Model Object
         */
        this.deleteField = function (field) { return __awaiter(_this, void 0, void 0, function () {
            function deleteRecursive(field, model) {
                var e_2, _a;
                return __awaiter(this, void 0, void 0, function () {
                    var modelKeys, modelKeys_1, modelKeys_1_1, fieldKey, _b, value, withoutProperty, _c, _d, _e, _f, _g, e_2_1, error_2;
                    var _h, _j;
                    return __generator(this, function (_k) {
                        switch (_k.label) {
                            case 0:
                                _k.trys.push([0, 16, , 17]);
                                modelKeys = Object.keys(model);
                                _k.label = 1;
                            case 1:
                                _k.trys.push([1, 9, 10, 15]);
                                modelKeys_1 = __asyncValues(modelKeys);
                                _k.label = 2;
                            case 2: return [4 /*yield*/, modelKeys_1.next()];
                            case 3:
                                if (!(modelKeys_1_1 = _k.sent(), !modelKeys_1_1.done)) return [3 /*break*/, 8];
                                fieldKey = modelKeys_1_1.value;
                                if (!(model[fieldKey][_constants_1.UI.NAME] === field)) return [3 /*break*/, 5];
                                return [4 /*yield*/, FormActions.deleteFieldPod(model[fieldKey])];
                            case 4:
                                _k.sent();
                                _b = fieldKey, value = model[_b], withoutProperty = __rest(model, [typeof _b === "symbol" ? _b : _b + ""]);
                                model = withoutProperty;
                                found = true;
                                return [3 /*break*/, 8];
                            case 5:
                                if (!model[fieldKey][_constants_1.UI.PARTS]) return [3 /*break*/, 7];
                                _c = [{}, model];
                                _h = {};
                                _d = fieldKey;
                                _e = [{}, model[fieldKey]];
                                _j = {};
                                _f = _constants_1.UI.PARTS;
                                _g = [{}];
                                return [4 /*yield*/, deleteRecursive(field, model[fieldKey][_constants_1.UI.PARTS])];
                            case 6:
                                model = __assign.apply(void 0, _c.concat([(_h[_d] = __assign.apply(void 0, _e.concat([(_j[_f] = __assign.apply(void 0, _g.concat([(_k.sent())])), _j)])), _h)]));
                                if (found) {
                                    return [3 /*break*/, 8];
                                }
                                _k.label = 7;
                            case 7: return [3 /*break*/, 2];
                            case 8: return [3 /*break*/, 15];
                            case 9:
                                e_2_1 = _k.sent();
                                e_2 = { error: e_2_1 };
                                return [3 /*break*/, 15];
                            case 10:
                                _k.trys.push([10, , 13, 14]);
                                if (!(modelKeys_1_1 && !modelKeys_1_1.done && (_a = modelKeys_1.return))) return [3 /*break*/, 12];
                                return [4 /*yield*/, _a.call(modelKeys_1)];
                            case 11:
                                _k.sent();
                                _k.label = 12;
                            case 12: return [3 /*break*/, 14];
                            case 13:
                                if (e_2) throw e_2.error;
                                return [7 /*endfinally*/];
                            case 14: return [7 /*endfinally*/];
                            case 15: return [2 /*return*/, model];
                            case 16:
                                error_2 = _k.sent();
                                throw error_2;
                            case 17: return [2 /*return*/];
                        }
                    });
                });
            }
            var partsObject, found, updatedModel, error_1;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        partsObject = this.formModel[_constants_1.UI.PARTS];
                        found = false;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, deleteRecursive(field, partsObject)
                            /**
                             * Update private formModel store
                             */
                        ];
                    case 2:
                        updatedModel = _b.sent();
                        /**
                         * Update private formModel store
                         */
                        this.formModel = __assign({}, this.formModel, (_a = {}, _a[_constants_1.UI.PARTS] = __assign({}, updatedModel), _a));
                        return [2 /*return*/, this.formModel];
                    case 3:
                        error_1 = _b.sent();
                        return [2 /*return*/, Error(error_1)];
                    case 4: return [2 /*return*/];
                }
            });
        }); };
        /**
         * Add new field into Form Model
         */
        this.addNewField = function (nodeName) {
            var _a;
            var found = false;
            var partsObject = _this.formModel[_constants_1.UI.PARTS];
            function findField(nodeName, model) {
                var _a, _b, _c, _d, _e;
                for (var fieldKey in model) {
                    var currentField = model[fieldKey];
                    if (currentField[_constants_1.UI.NAME] === nodeName && currentField['rdf:type'].includes('Multiple')) {
                        var copiedField = Object.keys(currentField[_constants_1.UI.CLONE_PARTS])[0];
                        var idLink = FormActions.getSubjectLinkId(currentField[_constants_1.UI.CLONE_PARTS][copiedField][_constants_1.UI.VALUE]);
                        var uniqueName = uuid_1.default();
                        model = __assign({}, model, (_a = {}, _a[fieldKey] = __assign({}, currentField, (_b = {}, _b[_constants_1.UI.PARTS] = __assign({}, currentField[_constants_1.UI.PARTS], (_c = {}, _c[uniqueName] = __assign({}, FormActions.cleanFieldNode(currentField[_constants_1.UI.CLONE_PARTS][copiedField]), (_d = {}, _d[_constants_1.UI.NAME] = uniqueName, _d[_constants_1.UI.VALUE] = idLink, _d)), _c)), _b)), _a));
                        found = true;
                        break;
                    }
                    else if (currentField[_constants_1.UI.PARTS]) {
                        model = __assign({}, model, (_e = {}, _e[fieldKey] = __assign({}, currentField, findField(nodeName, currentField[_constants_1.UI.PARTS])), _e));
                        if (found) {
                            break;
                        }
                    }
                }
                return model;
            }
            var updatedParts = findField(nodeName, partsObject);
            return __assign({}, _this.formModel, (_a = {}, _a[_constants_1.UI.PARTS] = __assign({}, updatedParts), _a));
        };
    }
    /**
     * Create random subject link for a node
     */
    FormActions.getSubjectLinkId = function (currentLink) {
        if (currentLink && currentLink.includes('#')) {
            var id = Date.now();
            return currentLink.split('#')[0] + "#" + id;
        }
    };
    /**
     * Clean field node into Form Model object
     */
    FormActions.cleanFieldNode = function (field) {
        var _a, _b, _c;
        var updatedField = field;
        if (updatedField && updatedField[_constants_1.UI.PARTS]) {
            for (var childKey in updatedField[_constants_1.UI.PARTS]) {
                updatedField = __assign({}, updatedField, (_a = {}, _a[_constants_1.UI.PARTS] = __assign({}, updatedField[_constants_1.UI.PARTS], (_b = {}, _b[childKey] = __assign({}, updatedField[_constants_1.UI.PARTS][childKey], FormActions.cleanFieldNode(updatedField[childKey])), _b)), _a));
            }
        }
        return __assign({}, updatedField, (_c = {}, _c[_constants_1.UI.VALUE] = '', _c[_constants_1.UI.OLDVALUE] = '', _c[_constants_1.UI.NAME] = uuid_1.default(), _c));
    };
    /**
     * Delete field from pod
     */
    FormActions.deleteFieldPod = function (field) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!field[_constants_1.UI.BASE]) return [3 /*break*/, 4];
                    if (!field[_constants_1.UI.PARENT_PROPERTY]) return [3 /*break*/, 2];
                    /**
                     * Delete field from  link data reference
                     */
                    return [4 /*yield*/, query_ldflex_1.default[field[_constants_1.UI.BASE]][field[_constants_1.UI.PARENT_PROPERTY]].delete(data_model_1.namedNode(field[_constants_1.UI.VALUE]))];
                case 1:
                    /**
                     * Delete field from  link data reference
                     */
                    _a.sent();
                    return [3 /*break*/, 4];
                case 2: return [4 /*yield*/, query_ldflex_1.default[field[_constants_1.UI.BASE]][field[_constants_1.UI.PROPERTY]].delete(field[_constants_1.UI.VALUE])];
                case 3:
                    _a.sent();
                    _a.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    }); };
    return FormActions;
}());
exports.FormActions = FormActions;
